File: frontend/src/satpass/tle.ts
/** Rebuild a TLE to the 69-character column layout SGP4 requires. */

function checksumDigit(line: string): string {
  let sum = 0;
  const body = line.slice(0, 68);
  for (const ch of body) {
    if (ch >= '0' && ch <= '9') sum += Number(ch);
    else if (ch === '-') sum += 1;
  }
  return String(sum % 10);
}

function padRight(s: string, n: number): string {
  const t = s.trim();
  return t.length >= n ? t.slice(0, n) : t.padEnd(n, ' ');
}

function padLeft(s: string, n: number): string {
  const t = s.trim();
  return t.length >= n ? t.slice(-n) : t.padStart(n, ' ');
}

function field10(raw: string): string {
  const t = raw.trim();
  if (t.startsWith('-') || t.startsWith('+')) return padLeft(t, 10);
  return padLeft(t, 10);
}

function field8(raw: string): string {
  return padLeft(raw.trim(), 8);
}

function parseLine1(line: string): string | null {
  const t = line.replace(/\r/g, '').trim();
  if (t.length === 69 && t.startsWith('1 ')) return t;
  const parts = t.split(/\s+/);
  if (parts[0] !== '1' || parts.length < 8) return null;

  let satnum = '';
  let cls = 'U';
  let idx = 1;
  if (/^\d{5}[A-Z]$/.test(parts[1])) {
    satnum = parts[1].slice(0, 5);
    cls = parts[1].slice(5);
    idx = 2;
  } else if (/^\d{5}$/.test(parts[1]) && /^[A-Z]$/.test(parts[2] || '')) {
    satnum = parts[1];
    cls = parts[2];
    idx = 3;
  } else {
    return null;
  }

  const intl = parts[idx++];
  const epoch = parts[idx++];
  const ndot = parts[idx++];
  const nddot = parts[idx++];
  const bstar = parts[idx++];
  const eph = parts[idx++];
  const tail = parts[idx] || '';
  if (!intl || !epoch || !ndot || !nddot || !bstar || eph == null || !tail) return null;
  if (!/^\d{5}\.\d+$/.test(epoch)) return null;

  const digits = tail.replace(/\D/g, '');
  const elnum = digits.length > 1 ? digits.slice(0, -1) : digits;

  const body =
    `1 ${satnum.padStart(5, '0')}${cls} ${padRight(intl, 8)} ${padLeft(epoch, 14)} ` +
    `${field10(ndot)} ${field8(nddot)} ${field8(bstar)} ${eph} ${padLeft(elnum, 4)}`;
  if (body.length !== 68) return null;
  return body + checksumDigit(body);
}

function parseLine2(line: string): string | null {
  const t = line.replace(/\r/g, '').trim();
  if (t.length === 69 && t.startsWith('2 ')) return t;
  const parts = t.split(/\s+/);
  if (parts[0] !== '2' || parts.length < 8) return null;
  const satnum = parts[1];
  const inc = parts[2];
  const raan = parts[3];
  const ecc = parts[4];
  const argp = parts[5];
  const ma = parts[6];
  let motion = parts[7];
  let rev = parts[8] || '';
  if (!rev && motion.length >= 16) {
    rev = motion.slice(11);
    motion = motion.slice(0, 11);
  }
  const revDigits = rev.replace(/\D/g, '');
  const revnum = revDigits.length > 5 ? revDigits.slice(0, 5) : revDigits.slice(0, 5);
  if (!/^\d{5}$/.test(satnum)) return null;

  const body =
    `2 ${satnum.padStart(5, '0')} ${padLeft(inc, 8)} ${padLeft(raan, 8)} ${padLeft(ecc, 7)} ` +
    `${padLeft(argp, 8)} ${padLeft(ma, 8)} ${padLeft(motion, 11)}${padLeft(revnum, 5)}`;
  if (body.length !== 68) return null;
  return body + checksumDigit(body);
}

/** Return column-aligned TLE lines, or the trimmed originals if they cannot be rebuilt. */
export function canonicalizeTle(line1: string, line2: string): { line1: string; line2: string } {
  return {
    line1: parseLine1(line1) ?? line1.trim(),
    line2: parseLine2(line2) ?? line2.trim(),
  };
}

File: backend/app/services/tle.py
"""Rebuild a TLE to the 69-character column layout SGP4 requires."""

from __future__ import annotations


def _checksum_digit(line: str) -> str:
    total = 0
    for ch in line[:68]:
        if ch.isdigit():
            total += int(ch)
        elif ch == "-":
            total += 1
    return str(total % 10)


def _pad_right(s: str, n: int) -> str:
    t = s.strip()
    return t[:n] if len(t) >= n else t.ljust(n)


def _pad_left(s: str, n: int) -> str:
    t = s.strip()
    return t[-n:] if len(t) >= n else t.rjust(n)


def _field10(raw: str) -> str:
    t = raw.strip()
    if t.startswith("-") or t.startswith("+"):
        return _pad_left(t, 10)
    return _pad_left(t, 10)


def _field8(raw: str) -> str:
    return _pad_left(raw.strip(), 8)


def canonicalize_line1(line: str) -> str:
    t = (line or "").replace("\r", "").strip()
    if len(t) == 69 and t.startswith("1 "):
        return t
    parts = t.split()
    if not parts or parts[0] != "1" or len(parts) < 8:
        return t
    idx = 1
    if len(parts[1]) == 6 and parts[1][:5].isdigit() and parts[1][5].isalpha():
        satnum, cls = parts[1][:5], parts[1][5]
        idx = 2
    elif parts[1].isdigit() and len(parts) > 2 and len(parts[2]) == 1 and parts[2].isalpha():
        satnum, cls = parts[1], parts[2]
        idx = 3
    else:
        return t
    try:
        intl, epoch, ndot, nddot, bstar, eph, tail = parts[idx : idx + 7]
    except ValueError:
        return t
    if not epoch or "." not in epoch:
        return t
    digits = "".join(ch for ch in tail if ch.isdigit())
    elnum = digits[:-1] if len(digits) > 1 else digits
    body = (
        f"1 {satnum.zfill(5)}{cls} {_pad_right(intl, 8)} {_pad_left(epoch, 14)} "
        f"{_field10(ndot)} {_field8(nddot)} {_field8(bstar)} {eph} {_pad_left(elnum, 4)}"
    )
    if len(body) != 68:
        return t
    return body + _checksum_digit(body)


def canonicalize_line2(line: str) -> str:
    t = (line or "").replace("\r", "").strip()
    if len(t) == 69 and t.startswith("2 "):
        return t
    parts = t.split()
    if not parts or parts[0] != "2" or len(parts) < 8:
        return t
    satnum, inc, raan, ecc, argp, ma, motion = parts[1:8]
    rev = parts[8] if len(parts) > 8 else ""
    if not rev and len(motion) >= 16:
        rev, motion = motion[11:], motion[:11]
    rev_digits = "".join(ch for ch in rev if ch.isdigit())[:5]
    if not satnum.isdigit():
        return t
    body = (
        f"2 {satnum.zfill(5)} {_pad_left(inc, 8)} {_pad_left(raan, 8)} {_pad_left(ecc, 7)} "
        f"{_pad_left(argp, 8)} {_pad_left(ma, 8)} {_pad_left(motion, 11)}{_pad_left(rev_digits, 5)}"
    )
    if len(body) != 68:
        return t
    return body + _checksum_digit(body)


def canonicalize_tle_line(line: str) -> str:
    raw = (line or "").strip()
    if raw.startswith("1 "):
        return canonicalize_line1(raw)
    if raw.startswith("2 "):
        return canonicalize_line2(raw)
    return raw
